aux_process.c

```diff
@@ -0,0 +1,1503 @@
+        goto error;
+    }
+
+    ptr = file_buf;
+
+    if (hand->output)
+        fprintf(hand->output, "\nchecksum for metadata file after applying each updater file is enable:\n");
+
+    for (i = 0; i < hand->num_mdfile_checksums; i++) {
+        UINT64DECODE(ptr, seq_nums[i]);
+        UINT32DECODE(ptr, hand->md_file_checksums[i]);
+
+        if (hand->output)
+            fprintf(hand->output, "\tupdater file %llu: checksum=%u\n", seq_nums[i], hand->md_file_checksums[i]);
+    }
+
+    if (fclose(hand->md_chksum_file) == EOF) {
+        fprintf(stderr, "updater file close failed\n");
+        goto error;
+    }
+
+    if (seq_nums)
+        free(seq_nums);
+
+    if (file_buf)
+        free(file_buf);
+
+    return 0;
+
+error:
+    if (seq_nums)
+        free(seq_nums);
+
+    if (file_buf)
+        free(file_buf);
+
+    return -1;
+}
+
+/*-------------------------------------------------------------------------
+ * Function: check_updater
+ *
+ * Purpose:  Wait for the updater file to appear, then apply it to the
+ *           metadata file.
+ *
+ * Return:   Success:    0
+ *
+ *           Failure:    -1
+ *-------------------------------------------------------------------------
+ */
+static int
+check_updater(handler_t *hand)
+{
+    int updater_exists;
+    int stop_update = 0;
+    char updater_name[FILE_NAME_LEN];
+    int i;
+
+    /* For testing: retrieve the metadata file checksum values if it exists */
+    if (hand->md_chksum_path && get_md_file_chksums(hand) < 0) {
+        fprintf(stderr, "failed to get the metadata file checksum with the name: %s\n", hand->md_chksum_path);
+        goto error;
+    }
+
+    /* Let it loop until an updater file appears */
+    for (i = 0; ; i++) {
+        sprintf(updater_name, "%s.%d", hand->updater_path, i);
+
+        updater_exists = -1;
+
+        do {
+            updater_exists = access(updater_name, F_OK);
+     
+            do_sleep(hand);
+        } while (updater_exists != 0);
+
+        stop_update = apply_updater(updater_name, hand);
+
+        if (stop_update)
+            break;
+        else if (stop_update < 0)
+            goto error;    
+    }
+
+    return 0;
+
+error:
+    return -1;
+}
+
+/*-------------------------------------------------------------------------
+ * Function: release_resources
+ *
+ * Purpose:  Free some memory
+ *
+ * Return:   Success:    0
+ *
+ *           Failure:    -1
+ *-------------------------------------------------------------------------
+ */
+static int
+release_resources(handler_t *hand)
+{
+    if (hand->log_file_path) {
+        free(hand->log_file_path);
+
+        if (fclose(hand->log_file) == EOF) {
+           fprintf(stderr, "log file close failed\n");
+           goto error;
+        }
+    }
+
+    if (hand->vfd_config)
+        free(hand->vfd_config);
+    
+    if (hand->updater_path)
+        free(hand->updater_path);
+
+    if (hand->md_file_path) 
+        free(hand->md_file_path);
+
+    if (hand->md_file_checksums)
```

Comment:
Oh, the md_file_checksums is released here. Adding a comment in the line when the buffer is allocated is helpful.

Reply:
Finished.